### PR TITLE
Add support for ```Variable``` symbol

### PR DESCRIPTION
**Pull request by Pranavchiku**
This PR generates ASR for multiple files inside ```scipy/linalg/src/id_dist/src```

### CODE REVIEW

**Review by czgdp1807 — COMMENTED** [review]
LGTM. But I will wait for https://github.com/lfortran/lfortran/pull/705 to go in first.

**Comment by Pranavchiku** [comment]
@czgdp1807 if you could merge this :)

**Comment by czgdp1807** [comment]
Tests fail. Update the reference tests. ASR.asdl changed after the sync with LPython.

**Comment by Pranavchiku** [comment]
I hope I didn't make it worse...

**Comment by czgdp1807** [comment]
Well seems like it is. 🥲.

**Comment by Pranavchiku** [comment]
> Well seems like it is. smiling_face_with_tear.

Yes it is ;-;-;

**Comment by konradha** [comment]
`[tests/variable1.f90](https://github.com/lfortran/lfortran/pull/703/files#diff-ccc2c6d2e484e5fcbdb1abcfd1b5aad95b8240b2f2797ef4f32f2c885efa8d48)` should not compile. `gfortran` throws with


```bash

    3 | complex x(m), y(n), p1, p2, p3, p4
      |          1
Error: Variable ‘m’ cannot appear in the expression at (1)
tests/variable1.f90:3:16:

    3 | complex x(m), y(n), p1, p2, p3, p4
      |                1
Error: Variable ‘n’ cannot appear in the expression at (1)
```

Would you also mind adding some more documentation @Pranavchiku ? I don't quite follow the changes :)

**Comment by Pranavchiku** [comment]
Originally it contained ```complex*16 x(m), y(n)```, but as ```complex*16``` throws error on ```llvm``` I converted that to ```complex x(m), y(n)```

**Comment by Pranavchiku** [comment]
```
subroutine variable01()
integer m,n
complex x, y, p1, p2, p3, p4
call matveca(x,m,y,p1,p2,p3,p4)
end
```

I guess this should work, I tried to check if ```matveca``` whose type is ```symbol``` compiles or not

**Comment by konradha** [comment]
`matveca` is a procedure -- @czgdp1807 is the typing correct then? It it legal to have a subroutine that's of `Variable` type? Just want to make sure that this fits into context :)

**Comment by czgdp1807** [comment]
Subroutine and Variable are separate symbols in ASR. So subroutine of Variable type doesn’t add up to much IMO.
If gfortran shows error for the test in the PR then please revert this PR. In a follow up PR and in other future PRs enable both gfortran and llvm option when you add integration tests so that we know that we are testing on correct things.

**Comment by konradha** [comment]
Quoting @Pranavchiku 

>  I tried to check if matveca symbol whose type is Variable compiles or not

which it should not be (if you look at the test case you can `matveca` being `call`ed). Maybe @Pranavchiku could provide some insight and/or fix; I don't think I should revert this change as I didn't click 'approve' here and I don't know the full story.

**Comment by Pranavchiku** [comment]
> `matveca` is a procedure (in the SciPy context) 

When I tried to debug it, it was actually showing me ```Variable``` type, hence I just added a provision for ```Variable``` over there...

**Comment by Pranavchiku** [comment]
>which it should not be (if you look at the test case you can matveca being called).

If it is so, ig than it must be because of ```create_implicit_interface_function``` in ```ast_common_visitor``` as ```matveca``` is an intrinsic function, which needs to be handled by creating implicit interface...

**Comment by konradha** [comment]
@Pranavchiku 

> I added the same code present for symbol Function in Variable to make it pass...

Yea well, but they're not the same.

@certik What do you think?

File: tests/variable1.f90 [hunk 1]
```diff
@@ -0,0 +1,5 @@
+subroutine variable01()
+integer m,n
+complex x(m), y(n), p1, p2, p3, p4
+call matveca(x,m,y,p1,p2,p3,p4)
+end
```
Comment (certik) [inline tests/variable1.f90:5]: This is not valid Fortran and should not compile.

Comment (certik) [reply tests/variable1.f90:5]: However it's a good test for a nice error message, so we'll reuse it for that. :)

**Comment by certik** [comment]
In general, this is an issue of our verify pass that must be a lot more strict to enforce the ASR semantics. Then it would be clear. Until then we have to be careful to construct ASR properly.

**Pull request merged**